src/coreclr/jit/optimizer.cpp

```diff
@@ -7770,338 +8539,90 @@ void Compiler::optOptimizeBools()
-                continue;
-            }
+            // The next block also needs to be a condition
 
-            // Make sure the types involved are of the same sizes
-            if (genTypeSize(c1->TypeGet()) != genTypeSize(c2->TypeGet()))
+            if (b2->bbJumpKind == BBJ_COND)
             {
-                continue;
-            }
-            if (genTypeSize(t1->TypeGet()) != genTypeSize(t2->TypeGet()))
-            {
-                continue;
-            }
-#ifdef TARGET_ARMARCH
-            // Skip the small operand which we cannot encode.
-            if (varTypeIsSmall(c1->TypeGet()))
-                continue;
-#endif
-            /* The second condition must not contain side effects */
-
-            if (c2->gtFlags & GTF_GLOB_EFFECT)
-            {
-                continue;
-            }
-
-            /* The second condition must not be too expensive */
-
-            gtPrepareCost(c2);
-
-            if (c2->GetCostEx() > 12)
-            {
-                continue;
-            }
-
-            genTreeOps foldOp;
-            genTreeOps cmpOp;
-            var_types  foldType = c1->TypeGet();
-            if (varTypeIsGC(foldType))
-            {
-                foldType = TYP_I_IMPL;
-            }
-
-            if (sameTarget)
-            {
-                /* Both conditions must be the same */
-
-                if (t1->gtOper != t2->gtOper)
+                if ((b1->bbJumpDest != b2->bbJumpDest) && (b1->bbJumpDest != b2->bbNext))
                 {
                     continue;
                 }
 
-                if (t1->gtOper == GT_EQ)
-                {
-                    /* t1:c1==0 t2:c2==0 ==> Branch to BX if either value is 0
-                       So we will branch to BX if (c1&c2)==0 */
+                // When it is conditional jumps
+                OptBoolsDsc optBoolsDsc;
+                optBoolsDsc.m_b1   = b1;
+                optBoolsDsc.m_b2   = b2;
+                optBoolsDsc.m_b3   = nullptr;
+                optBoolsDsc.m_comp = this;
 
-                    foldOp = GT_AND;
-                    cmpOp  = GT_EQ;
-                }
-                else
+                if (optBoolsDsc.optOptimizeBoolsCondBlock())
                 {
-                    /* t1:c1!=0 t2:c2!=0 ==> Branch to BX if either value is non-0
-                       So we will branch to BX if (c1|c2)!=0 */
-
-                    foldOp = GT_OR;
-                    cmpOp  = GT_NE;
+                    change = true;
                 }
             }
-            else
+            else if (b2->bbJumpKind == BBJ_RETURN)
             {
-                /* The b1 condition must be the reverse of the b2 condition */
+                // If there is no b1->bbJumpDest, we're done
 
-                if (t1->gtOper == t2->gtOper)
+                BasicBlock* b3 = b1->bbJumpDest;
+                if (b3 == nullptr)
                 {
-                    continue;
+                    break;
                 }
```

Comment:
This doesn't make sense: `b1->bbJumpDest` should *always* be non-null.

Reply:
Done. Removed.